src/test/java/jenkins/plugins/git/GitRepoSizeEstimatorTest.java

```diff
@@ -0,0 +1,125 @@
+     */
+    @Test
+    public void testSizeEstimationWithGitCache() throws Exception {
+        sampleRepo.init();
+        sampleRepo.git("checkout", "-b", "dev");
+        sampleRepo.write("file", "modified");
+        sampleRepo.git("commit", "--all", "--message=dev");
+        sampleRepo.git("tag", "lightweight");
+        sampleRepo.write("file", "modified2");
+        sampleRepo.git("commit", "--all", "--message=dev2");
+        sampleRepo.git("tag", "-a", "annotated", "-m", "annotated");
+        sampleRepo.write("file", "modified3");
+        sampleRepo.git("commit", "--all", "--message=dev3");
+        GitSCMSource source = new GitSCMSource(sampleRepo.toString());
+        TaskListener listener = StreamTaskListener.fromStderr();
+        // SCMHeadObserver.Collector.result is a TreeMap so order is predictable:
+        assertEquals("[]", source.fetch(listener).toString());
+        source.setTraits(Collections.<SCMSourceTrait>singletonList(new BranchDiscoveryTrait()));
+        assertEquals(GitBranchSCMHead_DEV_MASTER, source.fetch(listener).toString());
+
+        GitRepoSizeEstimator repoSizeEstimator = new GitRepoSizeEstimator(source);
+        /*
+        Since the size of repository is 21.785 KiBs, the estimator should suggest "jgit" as an implementation
+         */
+        assertThat(repoSizeEstimator.getGitTool(), containsString("git"));
+    }
+
+    @Test
+    public void testSizeEstimationWithGithubAPI() {
+        String remote = "https://github.com/rishabhBudhouliya/git-plugin.git";
+        GitRepoSizeEstimator sizeEstimator = new GitRepoSizeEstimator(remote);
+        assertThat(sizeEstimator.getGitTool(), is("git"));
+    }
+
+    @Test
+    public void testSizeEstimationWithBitbucketAPIs() {
+        String remote = "https://bitbucket.com/rishabhBudhouliya/git-plugin.git";
+        GitRepoSizeEstimator sizeEstimator = new GitRepoSizeEstimator(remote);
+        assertThat(sizeEstimator.getGitTool(), is("NONE"));
+    }
+
+    @org.jvnet.hudson.test.TestExtension
+    public static class TestExtensionGithub extends GitRepoSizeEstimator.RepositorySizeAPI {
+
+        @Override
+        public boolean acceptsRemote(String remote) {
+            return remote.contains("github");
+        }
+
+        @Override
+        public Long getSizeOfRepository(String remote) {
+            // from remote, remove .git and https://github.com
+            long mockedSize = 500;
+            return mockedSize;
+        }
+    }
+
+    @org.jvnet.hudson.test.TestExtension
+    public static class TestExtensionGitlab extends GitRepoSizeEstimator.RepositorySizeAPI {
+
+        @Override
+        public boolean acceptsRemote(String remote) {
+            return remote.contains("gitlab");
+        }
+
+        @Override
+        public Long getSizeOfRepository(String remote) {
+            // from remote, remove .git and https://github.com
+            long mockedSize = 1000;
+            return mockedSize;
```

Comment:
nit
```suggestion
            return 1000;
```

Reply:
what if instead 1000, you return 100000 and add another test for gitlab which should return "jgit"? that way you would be testing all the possibilities. When writing tests, always bare in mind all the possible returned values or you run the risk that a regression in your code will be never caught

Reply:
Agreed, I'm more test cases to cover as many possible test case scenarios.